### PR TITLE
Add deepspeed test to amd scheduled CI

### PR DESCRIPTION
**Pull request by echarlaix**

### CODE REVIEW

File: .github/workflows/self-scheduled-amd-caller.yml
Comment (echarlaix) [inline .github/workflows/self-scheduled-amd-caller.yml:8]: will remove this modification before merging (added to disable all the other AMD scheduled tests)

File: .github/workflows/self-nightly-scheduled.yml
Comment (echarlaix) [inline .github/workflows/self-nightly-scheduled.yml:215]: deepspeed build option `DS_BUILD_UTILS` is deprecated since `v0.10.0` (since https://github.com/microsoft/DeepSpeed/pull/3728)

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile:26]: This will (always) use the torch `2.0.1` which is not what we want. I suggest to 

- either based on `docker/transformers-pytorch-amd-gpu/Dockerfile` on current `main`
- or not even to build an AMD+deepespeed image: just build+install deepspeed at CI runtime
- otherwise, based on this image, but need to install the 

    - ARG PYTORCH='2.1.0'
    - ARG TORCH_VISION='0.16.0'
    - ARG TORCH_AUDIO='2.1.0'

**Review by ydshieh — COMMENTED** [review]
Thank you @echarlaix !

LGTM 🚀 , except one question in the docker file.

File: .github/workflows/self-nightly-scheduled.yml
Comment (ydshieh) [reply .github/workflows/self-nightly-scheduled.yml:215]: NICE!

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (echarlaix) [reply docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile]: Thanks @ydshieh that makes sense, just upgraded the torch version in https://github.com/huggingface/transformers/pull/27633/commits/f846b80bed3b75471edfde5ae3b617d0b235765a. 
I cannot use `docker/transformers-pytorch-amd-gpu/Dockerfile` at the moment as there is an incompatibility with ROCm and deepspeed for some reason (coming from the adam extension), but can modify it and change its parent image if needed

**Review by fxmarty — APPROVED** [review]
LGTM thank you!

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (echarlaix) [reply docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile:21]: I was getting similar issue than described in https://github.com/microsoft/DeepSpeed/issues/4513  (with `torch==2.1.0+rocm5.6` and `deepspeed==0.12.3`)

**Comment by fxmarty** [comment]
Running the tests, all tests passing on A100 pass as well on MI250. Two failing ones (that seem unrelated to deepspeed: https://github.com/huggingface/transformers/blob/e0d2e695827594c6a95a68612f32c65b2686985e/tests/extended/test_trainer_ext.py#L125, for both MI250/A100) are:

```
FAILED tests/extended/test_trainer_ext.py::TestTrainerExt::test_trainer_log_level_replica_1_low - AssertionError: 8 != 2
FAILED tests/extended/test_trainer_ext.py::TestTrainerExt::test_trainer_log_level_replica_2_high - AssertionError: 7 != 1
```

which I'll leave to an other PR.

One issue with the image `rocm/pytorch:rocm5.6_ubuntu20.04_py3.8_pytorch_2.0.1` is that it has `numba==0.49.0` (that is imported [here](https://github.com/huggingface/transformers/blob/e0d2e695827594c6a95a68612f32c65b2686985e/tests/models/wav2vec2/test_modeling_flax_wav2vec2.py#L67) through librosa) installed, which plays not nicely with transformers dependency resolution as it uses outdated `np.long`.

I have the warning `cpu_adam cuda is missing or is incompatible with installed torch, only cpu ops can be compiled!` as well with `rocm/pytorch:rocm5.6_ubuntu20.04_py3.8_pytorch_2.0.1` (for example `pytest tests/ -k "test_can_resume_training_errors_zero2_fp16" -s -vvvvv`), so I moved to use `rocm/dev-ubuntu-22.04:5.6` instead. Issue open in DeepSpeed about it: https://github.com/microsoft/DeepSpeed/issues/4768

File: .github/workflows/build-docker-images.yml
Comment (fxmarty) [reply .github/workflows/build-docker-images.yml:275]: @ydshieh is there something we need to do here?

**Comment by fxmarty** [comment]
@ydshieh Should be in good shape. There are many jobs though, but I guess it is the way transformers CI is designed? https://github.com/huggingface/transformers/actions/runs/7089122062/job/19293182925

**Comment by ydshieh** [comment]
Well, it's probably will affect tomorrow's scheduled CI.

**Comment by ydshieh** [comment]
Let's cancel it. I will show you how we usually do the experimentation tomorrow.

**Comment by fxmarty** [comment]
Sorry!

File: docker/transformers-pytorch-amd-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-pytorch-amd-gpu/Dockerfile:27]: Is this necessary? I mean, do you see any issue so adding this line to avoid it?

File: .github/workflows/build-docker-images.yml
Comment (ydshieh) [reply .github/workflows/build-docker-images.yml:275]: This is another issue that we can deal with outside this PR.

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (echarlaix) [reply docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile]: currently getting a `RuntimeError: Error building extension 'cpu_adam'`, I think we can go for images from  [rocm/pytorch](https://hub.docker.com/r/rocm/pytorch) for now

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile:34]: I would suggest to have

```
RUN DS_BUILD_CPU_ADAM=1 DS_BUILD_FUSED_ADAM=1 python3 -m pip install deepspeed --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check 2>&1
```
or whatever equivalent for deepspeed in ROCM if necessary.

File: .github/workflows/self-scheduled-amd.yml
Comment (ydshieh) [inline .github/workflows/self-scheduled-amd.yml:379]: maybe add

```
      # To avoid unknown test failures
      - name: Pre build DeepSpeed *again*
        working-directory: /workspace
        run: |
          python3 -m pip uninstall -y deepspeed
          DS_BUILD_CPU_ADAM=1 DS_BUILD_FUSED_ADAM=1 DS_BUILD_UTILS=1 python3 -m pip install deepspeed --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check
```
to be the same as in other workflow file.

File: .github/workflows/build-docker-images.yml
Comment (ydshieh) [reply .github/workflows/build-docker-images.yml:275]: But here we have to build the image manually.

**Comment by ydshieh** [comment]
Hi @fxmarty Let me know if you have any question or need help regarding my above comments.

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (fxmarty) [inline docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile:19]:
```suggestion
    hipblas-dev \
    rocblas-dev && \
```

**Comment by echarlaix** [comment]
Thanks @fxmarty @ydshieh for the updates while I was sick, let me push the new image manually, just launching all the tests locally to verify everything is working with the updated image before pushing

File: .github/workflows/self-scheduled-amd.yml
Comment (echarlaix) [reply .github/workflows/self-scheduled-amd.yml:379]: not sure to understand why we need to uninstall and reinstall deepspeed here, what issue does it solve ?

File: docker/transformers-pytorch-amd-gpu/Dockerfile
Comment (fxmarty) [reply docker/transformers-pytorch-amd-gpu/Dockerfile:27]: For the motivation: https://stackoverflow.com/questions/36996046/how-to-prevent-dockerfile-caching-git-clone

We do not want docker to cache the following `git clone`.

**Comment by ydshieh** [comment]
Hi @echarlaix You don't need to run all the tests. Just make sure 

- the image could build
- the deepspeed build step works 
- the deepspeed tests could be launched (we don't really care how many failing tests it has for now)

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (echarlaix) [reply docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile:35]: sure we can pre-compile deepspeed for this set of ops, was just wondering whether we can keep it in jit mode so that all the machine compatible ops can be dynamically build at runtime

Comment (ydshieh) [reply docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile:35]: jit mode will make some tests slower and potentially timeout, right?

File: .github/workflows/self-scheduled-amd.yml
Comment (ydshieh) [reply .github/workflows/self-scheduled-amd.yml:379]: I don't remember exactly, it has been one year or more ago. I can try to find from the history if you would like to have the information.

File: docker/transformers-pytorch-amd-gpu/Dockerfile
Comment (ydshieh) [reply docker/transformers-pytorch-amd-gpu/Dockerfile:27]: I am not familiar with this part. Is it relevant only if we build the image on the same machine multiple times?
So far we don't have such issue, but maybe that is because we build the image on Github Actions hosted machine rather than self-hosted VM.

Comment (fxmarty) [reply docker/transformers-pytorch-amd-gpu/Dockerfile:27]: I had the issue locally when using this docker image, where I would not pick up the latest transformers commit due to docker cache. I think it is useful to make sure we use the latest commit - even though in the CI this is not an issue.

Comment (ydshieh) [reply docker/transformers-pytorch-amd-gpu/Dockerfile:27]: ok。

So the issue occurs when you run the docker image, not at the time of docker build, right?

Comment (ydshieh) [reply docker/transformers-pytorch-amd-gpu/Dockerfile:27]: If this is more on when we use the image, I think it is expected. These images are not built for long-term usage: they are re-built on a **daily basis** to run the CI at **a specific and the same commit** .

Anyone wants to use the image locally is responsible to do `git fetch && git pull` (or `git checkout`).

If we accept this change (and IIUC), it means, on the CI (GitHub Actions),  each job may get different commits to run the test against, which is not what we want.

The above is just to explain the current behavior (before this change), not meaning we have issue, as on CI, we have

```
git fetch && git checkout ${{ github.sha }}
```
so we are safe.

Comment (fxmarty) [reply docker/transformers-pytorch-amd-gpu/Dockerfile:27]: > So the issue occurs when you run the docker image, not at the time of docker build, right?

No, `docker build` caches intermediate layers, one of them being `git clone https://github.com/huggingface/transformers && cd transformers && git checkout $REF`. If we use `REF=main` and actually want to use the latest commit, we need to invalidate the docker cache and this is what this `ADD` is doing.

Otherwise, docker cached intermediate layer is used and we may use an outdated commit compared to the latest commit available at build time.

Comment (ydshieh) [reply docker/transformers-pytorch-amd-gpu/Dockerfile:27]: OK, so during build time. Thanks for the detailed explanation.

File: .github/workflows/self-scheduled-amd.yml
Comment (echarlaix) [reply .github/workflows/self-scheduled-amd.yml:379]: in our case we don't need it at the moment, so does it work if we keep it that way ? If you want me to uninstall / reinstall it in the tests, I can directly update and use `huggingface/transformers-pytorch-amd-gpu` and just install deepspeed in the tests directly (to avoid this step)

File: docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile
Comment (ydshieh) [reply docker/transformers-pytorch-deepspeed-amd-gpu/Dockerfile]: Thanks @echarlaix . The most important is to do this again at CI time, as mentioned in the comment below.

(It may or may not relevant now, but I never checked again. I keep both to just avoid potential issue popping up)

File: .github/workflows/self-scheduled-amd.yml
Comment (echarlaix) [inline .github/workflows/self-scheduled-amd.yml:369]: To update `huggingface/transformers-pytorch-deepspeed-amd-gpu` manually this I would need to be added to https://hub.docker.com/u/huggingface cc @ydshieh

**Review by ydshieh — APPROVED** [review]
Thanks @fxmarty @echarlaix ! I pushed a commit to finalize it.

Ping @LysandreJik for a core maintainer's review.

**Review by LysandreJik — APPROVED** [review]
Thanks all!

**Pull request merged**
